cpp/src/column/column_device_view.cu

```diff
@@ -36,77 +36,74 @@ void column_device_view::destroy() {
-// memory ptr (d_ptr) to set any child object pointers.
-column_device_view::column_device_view( column_view source, void * h_ptr, void* d_ptr )
-    : detail::column_device_view_base{source.type(),       source.size(),
-                                      source.head(),       source.null_mask(),
-                                      source.null_count(), source.offset()},
-      _num_children{source.num_children()}
+namespace
 {
+
+// helper function for use by column_device_view and mutable_column_device_view constructors
+template<typename ColumnView, typename ColumnDeviceView>
+ColumnDeviceView* child_columns_to_device_array( ColumnView const& source, void* h_ptr, void* d_ptr )
+{
+  ColumnDeviceView* d_children = nullptr;
   size_type num_children = source.num_children();
   if( num_children > 0 )
   {
-    // The beginning of the memory must be the fixed-sized column_device_view
+    // The beginning of the memory must be the fixed-sized ColumnDeviceView
     // struct objects in order for d_children to be used as an array.
     // Therefore, any child data is assigned past the end of this array.
-    auto h_column = reinterpret_cast<column_device_view*>(h_ptr);
-    auto d_column = reinterpret_cast<column_device_view*>(d_ptr);
+    auto h_column = reinterpret_cast<ColumnDeviceView*>(h_ptr);
+    auto d_column = reinterpret_cast<ColumnDeviceView*>(d_ptr);
     auto h_end = reinterpret_cast<int8_t*>(h_column + num_children);
     auto d_end = reinterpret_cast<int8_t*>(d_column + num_children);
     d_children = d_column; // set member ptr to device memory
-    for( size_type idx=0; idx < _num_children; ++idx )
+    for( size_type idx=0; idx < num_children; ++idx )
     { // inplace-new each child into host memory
-      column_view child = source.child(idx);
-      new (h_column) column_device_view(child, h_end, d_end);
+      auto child = source.child(idx);
+      new (h_column) ColumnDeviceView(child, h_end, d_end);
       h_column++; // adv to next child
       // update the pointers for holding this child column's child data
-      auto col_child_data_size = extent(child) - sizeof(column_device_view);
+      auto col_child_data_size = ColumnDeviceView::extent(child) - sizeof(ColumnDeviceView);
       h_end += col_child_data_size;
       d_end += col_child_data_size;
     }
   }
+  return d_children;
 }
 
-// Construct a unique_ptr that invokes `destroy()` as it's deleter
-std::unique_ptr<column_device_view, std::function<void(column_device_view*)>>
-column_device_view::create(column_view source, cudaStream_t stream) {
-
+// helper function for column_device_view::create and mutable_column_device::create methods
+template<typename ColumnView, typename ColumnDeviceView>
+std::unique_ptr<ColumnDeviceView, std::function<void(ColumnDeviceView*)>>
+create_device_view_from_view( ColumnView const& source, cudaStream_t stream )
+{
   size_type num_children = source.num_children();
```

Comment:
I feel like this function is getting intimate enough with the structure of the class that it really should be a class member.

Reply:
The templating allows for the common code to exist for both `column_device_view` and `mutable_column_device_view` and requires such compile time data as sizeof() each class. I feel it is well hidden (intimate?) defined here in an unnamed namespace instead of declared in a header file.